### PR TITLE
Issue #2810: Feature to have JavadocMethod validate parameter order 

### LINKED ISSUE #2810: Feature to have JavadocMethod validate parameter order
TestClass.java

```
package test;

public class TestClass {
    /**
     * Description.
     * 
     * @param p2
     *            Parameter 2 desc.
     * @param p1
     *            Parameter 1 desc.
     * @param <T>
     *            Parameter T desc.
     * @param p3
     *            Parameter 3 desc.
     */
    public <T> void test(String p1, String p2, T p3) {
    }
}
```

config.xml

```
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="ASCII" />
    <property name="severity" value="error" />
    <property name="fileExtensions" value="java" />

    <module name="TreeWalker">
        <module name="JavadocMethod">
            <property name="scope" value="public" />
            <property name="allowedAnnotations" value="Override" />
        </module>
    </module>
</module>
```

output from: `java -jar checkstyle-6.14.1-all.jar -c config.xml TestClass.java`

```
Starting audit...
Audit done.
```

To add less confusion between the javadoc order and the method when reading, I would like the at parameters to be validated in the javadoc so they both agree in terms of order. This will give new readers less headaches as they try to read and understand the method and what it's parameters mean.

With this new feature, I am expecting the javadocs for parameters 1 and T to produce errors that they are out of order in my example. If 2 and/or 3 produce errors that is fine, but I expect the main ones that go backwards in the list to definitely be there.
The preferred order, imo, should be T, 1, 2, 3 like they are defined in the method definition.
##

### PR DESCRIPTION
**Pull request by Gaurav-Punjabi**
Fixes #2810 

To add order check-in `JavadocMethod`. As we needed the original order of the param, hence I had to maintain a copy of the original param list through which we search the `position` of parameter with `@param` tag.

### CODE REVIEW

**Comment by Gaurav-Punjabi** [comment]
@romani @rnveach Please review. I have referenced the issue in my PR description. Then why is my Travis build still failing?

**Comment by rnveach** [comment]
>  why is my Travis build still failing

Please read: https://travis-ci.org/github/checkstyle/checkstyle/jobs/666377544#L286
> You are providing a PR for an Issue that is not approved yet,
please ask admins to approve your Issue first

Wercker failure is with orekit:
> [INFO] There are 321 errors reported by Checkstyle 8.31-SNAPSHOT with /pipeline/source/.ci-temp/Orekit/checkstyle.xml ruleset.
[ERROR] src/main/java/org/orekit/gnss/Frequency.java:[128,8] (javadoc) JavadocMethod: Incorrect order of @param tag for 'name'.

If issue is approved and we enforce this sorting by default, all failures in other projects will have to be resolved for CI to pass. This will means we either need fix our code if the violations are wrong or need to send PRs to those projects telling them why this change is coming and that they need to accept the changes or tell us why they won't.

If issue is approved, does it make sense we enforce parameter order by default?

**Comment by rnveach** [comment]
@romani ping

**Comment by Gaurav-Punjabi** [comment]
> You are providing a PR for an Issue that is not approved yet,
> please ask admins to approve your Issue first

Ohh my bad.

> If issue is approved, does it make sense we enforce parameter order by default?

What if we provide a property `checkOrder` and keep it false by default. Due to this other projects would be affected.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/AnnotationUseStyleCheck.java [hunk 1]
```diff
@@ -311,9 +311,9 @@ public void setClosingParens(final String parens) {
 
     /**
      * Retrieves an {@link Enum Enum} type from a @{link String String}.
-     * @param <T> the enum type
      * @param enumClass the enum class
      * @param value the string representing the enum
+     * @param <T> the enum type
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/AnnotationUseStyleCheck.java]: https://github.com/checkstyle/checkstyle/issues/2810
> The preferred order, imo, should be T, 1, 2, 3 like they are defined in the method definition.

This change goes against what was described in issue.

Comment (Gaurav-Punjabi) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/AnnotationUseStyleCheck.java]: Oh yeah, I'll fix this one.

**Comment by Gaurav-Punjabi** [comment]
> > What if we provide a property checkOrder and keep it false by default. Due to this other projects would be affected.
> 
> I am fine with property, but not name. It should be clear what the property is doing by its name. Since it is enforcing order of parameters it should be something like `validateParameterOrder`.

Alright I'll add a property with this name.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/AnnotationUseStyleCheck.java (hunk 1 above)
Comment (Gaurav-Punjabi) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/AnnotationUseStyleCheck.java]: Done 👍

**Comment by Gaurav-Punjabi** [comment]
@rnveach Added a property `validateParameterOrder` which is false by default. So it won't affect the current projects. Please review.

**Comment by rnveach** [comment]
@Gaurav-Punjabi pitest must be satisfied next. If your not familar with pitest, I recommend looking at main pitest issue as it says how to work through it.

**Comment by Gaurav-Punjabi** [comment]
> @Gaurav-Punjabi pitest must be satisfied next. If your not familar with pitest, I recommend looking at main pitest issue as it says how to work through it.

Yeah, I'm familiar with pitest. I'll kill the surviving mutations and ping you once it's done.

**Comment by Gaurav-Punjabi** [comment]
@rnveach Fixed it. No mutations are surviving now. Please review.

File: src/xdocs/config_javadoc.xml [hunk 2]
```diff
@@ -649,6 +649,13 @@ public int checkReturnTag(final int aTagIndex,
               <td><code>false</code></td>
               <td>6.0</td>
             </tr>
+            <tr>
+              <td>validateParameterOrder</td>
+              <td>Control whether to validate order of <code>param</code> tags.</td>
```
Comment (rnveach) [inline src/xdocs/config_javadoc.xml:654]: We need to expand on correct order. They should be ordered as they appear in the code.

Comment (rnveach) [inline src/xdocs/config_javadoc.xml:654]: We need a config example of new property in use.

File: src/xdocs/config_javadoc.xml [hunk 3]
```diff
@@ -649,6 +649,13 @@ public int checkReturnTag(final int aTagIndex,
...
+              <td><a href="property_types.html#boolean">Boolean</a></td>
+              <td><code>false</code></td>
+              <td>8.4</td>
```
Comment (rnveach) [inline src/xdocs/config_javadoc.xml:657]: This is being released in 8.4? Last I checked we are working towards 8.31 .

Comment (Gaurav-Punjabi) [reply src/xdocs/config_javadoc.xml:657]: I saw that the last release of check style on GitHub was 8.4. So I wrote 8.4. What should I write instead ?

File: src/xdocs/config_javadoc.xml (hunk 2 above)
Comment (Gaurav-Punjabi) [reply src/xdocs/config_javadoc.xml:654]: Alright, will add one.

Comment (Gaurav-Punjabi) [reply src/xdocs/config_javadoc.xml:654]: Okay 👍

File: src/xdocs/config_javadoc.xml (hunk 3 above)
Comment (rnveach) [reply src/xdocs/config_javadoc.xml:657]: https://github.com/checkstyle/checkstyle/releases
We just release 8.31 . So next release is 8.32 . I am not sure where you saw 8.4 .

**Comment by Gaurav-Punjabi** [comment]
@rnveach Made these changes. Please review.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java [hunk 4]
```diff
@@ -811,6 +863,46 @@ private void checkParamTags(final List<JavadocTag> tags,
...
+
+            final boolean isTypeParam = CommonUtil.startsWithChar(arg1, '<')
+                    && CommonUtil.endsWithChar(arg1, '>');
+            if (isTypeParam) {
+                actualPosition = indexOfTypeParameter(allTypeParams,
+                        arg1.substring(1, arg1.length() - 1));
+            }
+
+            // Handle incorrect JavadocTag order
+            if (validateParameterOrder && actualPosition != paramPosition) {
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java:898]: We have another option to allow skipping documentation of some parameters, `allowMissingParamTags`. Won't using `!=` cause issues with that instead of using `>` or `<`? Please add a test with `allowMissingParamTags` turned on and some examples of out of order parameters and in-order parameters.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java [hunk 5]
```diff
@@ -811,6 +863,46 @@ private void checkParamTags(final List<JavadocTag> tags,
...
+
+            final String arg1 = tag.getFirstArg();
+            int actualPosition = indexOfParam(allParams, arg1) + noOfTypeParams;
+
+            final boolean isTypeParam = CommonUtil.startsWithChar(arg1, '<')
+                    && CommonUtil.endsWithChar(arg1, '>');
+            if (isTypeParam) {
+                actualPosition = indexOfTypeParameter(allTypeParams,
+                        arg1.substring(1, arg1.length() - 1));
+            }
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java:895]: why not add
````
} else {
  actual = indexOfParam(....)
}
````

Instead of executing it all the time?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java [hunk 6]
```diff
@@ -782,15 +832,17 @@ private void checkParamTags(final List<JavadocTag> tags,
             final String arg1 = tag.getFirstArg();
             boolean found = removeMatchingParam(params, arg1);
 
-            if (CommonUtil.startsWithChar(arg1, '<') && CommonUtil.endsWithChar(arg1, '>')) {
+            final boolean isTypeParam = CommonUtil.startsWithChar(arg1, '<')
+                    && CommonUtil.endsWithChar(arg1, '>');
+            if (isTypeParam) {
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java:837]: Why is this change needed? I am not seeing us use the new variable anywhere?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java [hunk 7]
```diff
@@ -767,6 +814,9 @@ private void checkParamTags(final List<JavadocTag> tags,
         final List<DetailAST> params = getParameters(parent);
         final List<DetailAST> typeParams = CheckUtil
                 .getTypeParameters(parent);
+        // Validating the order of these parameters
+        checkParamTagsOrder(new ArrayList<>(tags), new ArrayList<>(params),
+                new ArrayList<>(typeParams));
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java:819]: is there a reason we can't send the list as is without re-wrapping around a new instance?

Comment (Gaurav-Punjabi) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java:819]: Because we are deleting the parameters from the list. Deletion changes the order of the parameters. Hence we created a copy of the original list, so it maintains the original order of all the parameters.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java (hunk 6 above)
Comment (Gaurav-Punjabi) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java:837]: Initially, I was using isTypeParam twice, but then I modified that code into one condition. Yes, now we don't need this variable. I'll write the condition inside `if`.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java (hunk 5 above)
Comment (Gaurav-Punjabi) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java:895]: Yeah, I think this would be better. I'll make this change.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java [hunk 8]
```diff
@@ -836,6 +928,48 @@ private static boolean searchMatchingTypeParameter(List<DetailAST> typeParams,
...
+     * @param requiredTypeName name of required type
+     * @return index of parameter if found else -1
+     */
+    private static int indexOfTypeParameter(List<DetailAST> typeParams,
+                                           String requiredTypeName) {
+        final Iterator<DetailAST> typeParamsIt = typeParams.iterator();
+        int paramIndex = -1;
+        int index = 0;
+        while (typeParamsIt.hasNext()) {
+            final DetailAST typeParam = typeParamsIt.next();
```
[the remark this answers is not in the document]
Comment (Gaurav-Punjabi) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java:943]: No reason. I'll refactor this method into a for each loop.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java [hunk 9]
```diff
@@ -811,6 +863,46 @@ private void checkParamTags(final List<JavadocTag> tags,
...
+     * @param allTypeParams list of type parameters
+     */
+    private void checkParamTagsOrder(final List<JavadocTag> tags,
+                                     final List<DetailAST> allParams,
+                                     final List<DetailAST> allTypeParams) {
+        // To store the actual position of the parameter, which will be
+        // later used for validating order of the params
+        int paramPosition = 0;
+        // Will the total no of type param present to validate order of parameters.
+        final int noOfTypeParams = allTypeParams.size();
```
[the remark this answers is not in the document]
Comment (Gaurav-Punjabi) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java:879]: Okay

File: src/xdocs/config_javadoc.xml [hunk 10]
```diff
@@ -778,6 +788,24 @@ public void doSomething9(File file) throws IOException {
...
+&lt;/module&gt;
+        </source>
+        <p>Example: </p>
+        <source>
+/**
+ * @param b // violation, b should be the second parameter as show in the code below
+ * @param a // violation, a should be the first parameter as shown in the code below
+ * @param c // OK
+ */
+public void test(int a, int b, int c) {}
```
[the remark this answers is not in the document]
Comment (Gaurav-Punjabi) [reply src/xdocs/config_javadoc.xml:807]: Alright, will add this.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java (hunk 4 above)
Comment (Gaurav-Punjabi) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java:898]: Yes,`allowMissingParamTags` will cause issues with the order. Can we skip checking the order of the parameters if `allowMissingParamTags' is true?

**Comment by Gaurav-Punjabi** [comment]
@romani Here's the regression report, 
1. [Default config](https://gaurav-punjabi.github.io/checkstyle/issue-2810/default-config/index.html)
2. [Config with validateParameterOrder set](https://gaurav-punjabi.github.io/checkstyle/issue-2810/paramter-order/index.html)
3. [Config with valiateParameterOrder and allowMissingParam set](https://gaurav-punjabi.github.io/checkstyle/issue-2810/missing-param/index.html)

File: src/xdocs/config_javadoc.xml (hunk 10 above)
[the remark this answers is not in the document]
Comment (Gaurav-Punjabi) [reply src/xdocs/config_javadoc.xml]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (Gaurav-Punjabi) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java:891]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (Gaurav-Punjabi) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java (hunk 5 above)
Comment (Gaurav-Punjabi) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java:906]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java (hunk 6 above)
Comment (Gaurav-Punjabi) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java:848]: Done

**Comment by romani** [comment]
@Gaurav-Punjabi , sorry for delay in reply, I assigned PR to me.

while I am looking into it, please rebase PR to resolve conflicts.

**Comment by romani** [comment]
@Gaurav-Punjabi , ping.

**Comment by romani** [comment]
@Gaurav-Punjabi , ping.
please finish PR.

**Comment by romani** [comment]
I voting to close this PR. It is abandoned, issue us not approved, implementation is not based on AST , so unlikely to be robust.

@rnveach .

**Comment by rnveach** [comment]
I do not have the time to take this up.

**Comment by romani** [comment]
new check should be created https://github.com/checkstyle/checkstyle/issues/8990
but we should agree on design before implementation.
